### PR TITLE
support utf-8 type in c_to_ml_type

### PR DESCRIPTION
**Pull request by martindemello**
I think this is fine - not very familiar with C interop but lablgtk appears to be using `caml_copy_string` too

### CODE REVIEW

**Comment by btj** [comment]
The return value of [`gtk_button_get_label`](https://docs.gtk.org/gtk4/method.Button.get_label.html) can be `NULL`. Indeed, the XML file has `allow-none="1"`:
```xml
      <method name="get_label" c:identifier="gtk_button_get_label" glib:get-property="label">
        <attribute name="org.gtk.Method.get_property" value="label"/>
        <return-value transfer-ownership="none" allow-none="1">
          <type name="utf8"/>
        </return-value>
      </method>
```
`caml_copy_string` will probably crash on `NULL`. I guess we should return a `string option` in this case.

**Comment by btj** [comment]
Note: [`gtk_entry_buffer_get_text`](https://docs.gtk.org/gtk4/method.EntryBuffer.get_text.html), which is a more compelling use case for a getter anyway, does not return `NULL`. So if you prefer, it seems fine to skip methods whose return value allows null for now.

**Comment by btj** [comment]
Also, we should be sure to skip methods whose return value does not have `transfer-ownership="none"`, because that might mean it's our job to deallocate the C string.

**Comment by martindemello** [comment]
okay, confined support to strings without `allow-none` for now. we should likely be checking `allow-none` for other types as well.

File: generate_bindings.py [hunk 1]
```diff
@@ -190,6 +191,39 @@ def drop_first(self):
...
+        typ = elem.find(t_type)
+        array = False
+        if typ is None:
+            array_elem = elem.find(t_array)
+            assert array_elem is not None, f"Unknown type {elem}"
+            array = True
+            typ = array_elem.find(t_type)
+            assert typ is not None, f"No type found for {elem}"
+        allow_none = elem.get('allow-none', "0") == "1"
+        transfer_ownership = elem.get('transfer-ownership', None)
```
Comment (btj) [inline generate_bindings.py:224]: Not a `bool`?

File: generate_bindings.py [hunk 2]
```diff
@@ -190,6 +191,39 @@ def drop_first(self):
...
+        direction = elem.get('direction', None)
```
Comment (btj) [inline generate_bindings.py:225]: Type should be `Optional[str]`?

**Review by btj — CHANGES_REQUESTED** [review]
Looks good! Just a few seemingly incorrect type annotations

File: generate_bindings.py (hunk 1 above)
Comment (martindemello) [reply generate_bindings.py:224]: oops, right. fixed the annotation

File: generate_bindings.py (hunk 2 above)
Comment (martindemello) [reply generate_bindings.py:225]: done

File: ml_gobject0.c [hunk 3]
```diff
@@ -40,6 +41,19 @@ value Val_GObject(GObject *obj) {
   return result;
 }
 
+value Val_string_option(char *s) {
+  CAMLparam0();
+  if (s) {
+    CAMLlocal2(ml_s, result);
```
Comment (btj) [inline ml_gobject0.c]: > Rule 2  Local variables of type value must be declared with one of the CAMLlocal macros. Arrays of values are declared with CAMLlocalN. These macros must be used at the beginning of the function, not in a nested block.

File: ml_gobject0.c [hunk 4]
```diff
@@ -40,6 +41,19 @@ value Val_GObject(GObject *obj) {
...
+    ml_s = caml_copy_string(s);
+    result = caml_alloc(1, 0);
+    Store_field(result, 0, ml_s);
```
Comment (btj) [inline ml_gobject0.c]: Use `caml_alloc_some`?

**Review by btj — CHANGES_REQUESTED** [review]
Looks good, except for some issues with the C code

File: ml_gobject0.c (hunk 3 above)
Comment (martindemello) [reply ml_gobject0.c]: thanks, done

File: ml_gobject0.c (hunk 4 above)
Comment (martindemello) [reply ml_gobject0.c]: done. wasn't sure if we wanted to drop support for pre-4.12.

**Comment by btj** [comment]
I'm afraid two more changes are necessary:
- Please make the parameter of `Val_string_option` of type `const char *` to avoid the warnings
- If and only if `platform.system() == 'Darwin'`, include header `gio/osxappinfo.h`

**Comment by btj** [comment]
Sorry, that should be `<gio/gosxappinfo.h>`

**Comment by martindemello** [comment]
done, the UI isn't letting me mark those as resolved

File: generate_bindings.py [hunk 5]
```diff
@@ -327,6 +328,9 @@ def print_class(self, class_):
 #include "ml_gobject0.h"\
 '''
 
+if platform.system == 'Darwin':
```
Comment (btj) [inline generate_bindings.py]: Should be `system()`

**Review by btj — APPROVED** [review]

File: generate_bindings.py (hunk 5 above)
Comment (martindemello) [reply generate_bindings.py]: doh :)

**Pull request merged**